iroh-net/src/magicsock/udp_conn.rs

```diff
@@ -29,41 +28,88 @@ impl UdpConn {
 
     pub(super) fn bind(addr: SocketAddr) -> anyhow::Result<Self> {
         let sock = bind(addr)?;
-        let state = quinn_udp::UdpSocketState::new(quinn_udp::UdpSockRef::from(&sock))?;
+        let state = sock.with_socket(|socket| {
+            quinn_udp::UdpSocketState::new(quinn_udp::UdpSockRef::from(socket))
+        })??;
+
         Ok(Self {
             io: Arc::new(sock),
-            inner: Arc::new(state),
+            inner: RwLock::new(state),
         })
     }
 
+    pub(super) fn rebind(&self) -> anyhow::Result<()> {
+        // Rebind underlying socket
+        self.io.rebind()?;
+
+        // update socket state
+        let new_state = self.io.with_socket(|socket| {
+            quinn_udp::UdpSocketState::new(quinn_udp::UdpSockRef::from(socket))
+        })??;
```

Comment:
If this fails the UdpConn is left in an unknown and unusable state right?  But this takes `&self` so there is no real indicator that this is broken afterwards.

Reply:
there is? all the methods will return an error if this happens